concurrent/src/main/java/io/airlift/concurrent/ResourceLoader.java

```diff
@@ -0,0 +1,224 @@
+    private final Consumer<Exception> exceptionHandler;
+    private final AtomicReference<Runnable> stopRefresh = new AtomicReference<>();
+
+    @GuardedBy("this")
+    private T value;
+
+    private static <T> ResourceLoader<T> createResourceLoader(
+            File file,
+            FileLoader<T> fileLoader,
+            Duration refreshPeriod,
+            SimpleRecurringTaskRunner recurringTaskRunner,
+            Consumer<Exception> exceptionHandler)
+            throws Exception
+    {
+        requireNonNull(recurringTaskRunner, "recurringTaskRunner is null");
+        ResourceLoader<T> resourceLoader = new ResourceLoader<>(
+                fileLoader,
+                file,
+                refreshPeriod,
+                exceptionHandler);
+        resourceLoader.start(recurringTaskRunner);
+        return resourceLoader;
+    }
+
+    private ResourceLoader(FileLoader<T> fileLoader, File file, Duration refreshPeriod, Consumer<Exception> exceptionHandler)
+            throws Exception
+    {
+        this.fileLoader = requireNonNull(fileLoader, "fileLoader is null");
+        this.file = requireNonNull(file, "file is null");
+        this.refreshPeriod = requireNonNull(refreshPeriod, "refreshPeriod is null");
+        this.exceptionHandler = requireNonNull(exceptionHandler, "exceptionHandler is null");
+        refreshNow();
+    }
+
+    private void start(SimpleRecurringTaskRunner recurringTaskRunner)
+    {
+        stopRefresh.set(recurringTaskRunner.scheduleTask(this::refreshInternal, refreshPeriod));
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        Runnable stopRefresh = this.stopRefresh.getAndSet(null);
+        if (stopRefresh != null) {
+            stopRefresh.run();
+        }
+    }
+
+    public synchronized T getValue()
+    {
+        return value;
+    }
+
+    public synchronized void refreshNow()
+            throws Exception
+    {
+        value = fileLoader.loadFile(file);
+    }
+
+    private void refreshInternal()
+    {
+        try {
+            refreshNow();
+        }
+        catch (Exception e) {
+            exceptionHandler.accept(e);
+        }
+    }
+
+    public interface FileLoader<T>
+    {
+        T loadFile(File file)
+                throws Exception;
+    }
+
+    public interface SimpleRecurringTaskRunner
+    {
+        Runnable scheduleTask(Runnable runnable, Duration fixedDelay);
+    }
+
+    private static class InternalScheduledRecurringTaskRunner
+            implements SimpleRecurringTaskRunner
+    {
+        private final String name;
+
+        private InternalScheduledRecurringTaskRunner(String name)
+        {
+            this.name = requireNonNull(name, "name is null");
+        }
+
+        @Override
+        public Runnable scheduleTask(Runnable runnable, Duration fixedDelay)
+        {
+            ScheduledExecutorService executor = newSingleThreadScheduledExecutor(daemonThreadsNamed(name + "-background-loader"));
```

Comment:
This might be too simple. If we have 50 refreshable configurations, we will have 50 threads idling.
Maybe a cached thread pool?